File: apps/desktop/src/db/models/user/user.interface.ts
/**
 * User Interface Definitions
 *
 * Defines the UserInterface for user information, subscription status,
 * feature permissions, and application state.
 *
 * @requirements 2.1
 */

import type { UUID, ISODateString } from "../shared";

/**
 * User subscription plan type
 */
export type UserPlan = "free" | "premium";

/**
 * Token validation status
 */
export type TokenStatus = "valid" | "invalid" | "unchecked";

/**
 * User feature permissions
 * Controls access to premium features based on subscription
 */
export interface UserFeatures {
  /** Whether user can use all scenes */
  canUseAllScenes?: boolean;
  /** Whether user can export to PDF */
  canExportPDF?: boolean;
  /** Whether user can use cloud sync */
  canUseCloudSync?: boolean;
  /** Whether to show ads (free tier) */
  showAds?: boolean;
  /** Reminder/ad popup interval in seconds */
  reminderInterval?: number;
  /** Flexible feature flags for future expansion */
  featureFlags?: Record<string, boolean>;
}

/**
 * User application state
 * Stores the user's current editing context for session restoration
 */
export interface UserState {
  /** Last opened location/route */
  lastLocation: string;
  /** Currently open project ID */
  currentProject: string;
  /** Currently editing chapter ID (legacy compatibility) */
  currentChapter: string;
  /** Currently editing scene ID (legacy compatibility) */
  currentScene: string;
  /** Currently editing title */
  currentTitle: string;
  /** Current typing text */
  currentTyping: string;
  /** Last cloud save timestamp */
  lastCloudSave: string;
  /** Last local save timestamp */
  lastLocalSave: string;
  /** Whether user is logged in */
  isUserLoggedIn: boolean;
}

/**
 * User settings/preferences
 */
export interface UserSettings {
  /** UI theme (light/dark) */
  theme: string;
  /** Application language (zh/en) */
  language: string;
  /** Enable autosave */
  autosave?: boolean;
  /** Enable spell check */
  spellCheck?: boolean;
  /** Remember last editing location */
  lastLocation?: boolean;
  /** Editor font size */
  fontSize: string;
}

/**
 * Database version info for user data
 */
export interface UserDBVersion {
  /** Version record ID */
  id: string;
  /** Database version string */
  version: string;
  /** Last update timestamp */
  updatedAt: string;
  /** Migration notes */
  migrationNotes?: string;
}

/**
 * Main User interface
 * Contains all user-related data including profile, subscription, and settings
 */
export interface UserInterface {
  /** Unique user identifier (UUID) */
  id: UUID;

  /** Login username */
  username: string;

  /** Display name (optional) */
  displayName?: string;

  /** Avatar URL (optional) */
  avatar?: string;

  /** User email (optional) */
  email?: string;

  /** Last login timestamp */
  lastLogin: ISODateString;

  /** Account creation timestamp */
  createDate: ISODateString;

  // Subscription information
  /** User subscription plan */
  plan: UserPlan;

  /** Subscription start date */
  planStartDate?: ISODateString;

  /** Subscription expiration date (premium) */
  planExpiresAt?: ISODateString;

  /** Trial expiration date */
  trialExpiresAt?: ISODateString;

  // Token and authentication
  /** Authentication token */
  token?: string;

  /** Token validation status */
  tokenStatus?: TokenStatus;

  /** Last token validation timestamp */
  lastTokenCheck?: ISODateString;

  /** Server message (subscription notices, etc.) */
  serverMessage?: string;

  // Nested objects
  /** Feature permissions */
  features?: UserFeatures;

  /** Application state */
  state?: UserState;

  /** User settings/preferences */
  settings?: UserSettings;

  /** Database version info */
  dbVersion?: UserDBVersion;
}

/**
 * User creation input type
 * Used when creating new users
 * id, createDate, and lastLogin are auto-generated
 */
export interface UserCreateInput {
  username: string;
  displayName?: string;
  avatar?: string;
  email?: string;
  plan?: UserPlan;
  features?: UserFeatures;
  settings?: UserSettings;
}

/**
 * User update input type
 * Used when updating existing users
 * Only mutable fields can be updated
 */
export interface UserUpdateInput {
  username?: string;
  displayName?: string;
  avatar?: string;
  email?: string;
  lastLogin?: ISODateString;
  plan?: UserPlan;
  planStartDate?: ISODateString;
  planExpiresAt?: ISODateString;
  trialExpiresAt?: ISODateString;
  token?: string;
  tokenStatus?: TokenStatus;
  lastTokenCheck?: ISODateString;
  serverMessage?: string;
  features?: UserFeatures;
  state?: UserState;
  settings?: UserSettings;
  dbVersion?: UserDBVersion;
}

File: packages/rust-core/src/types/user/user_interface.rs
//! User DTO 接口定义
//!
//! 定义用户相关的数据传输对象（DTO）

use super::user_entity::UserPlan;
use serde::{Deserialize, Serialize};

// ============================================================================
// 嵌套类型
// ============================================================================

/// 用户功能权限
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct UserFeatures {
    pub can_use_all_scenes: Option<bool>,
    pub can_export_pdf: Option<bool>,
    pub can_use_cloud_sync: Option<bool>,
    pub show_ads: Option<bool>,
    pub reminder_interval: Option<i32>,
}

/// 用户应用状态
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct UserState {
    pub last_location: Option<String>,
    pub current_project: Option<String>,
    pub current_chapter: Option<String>,
    pub current_scene: Option<String>,
    pub current_title: Option<String>,
    pub current_typing: Option<String>,
    pub last_cloud_save: Option<String>,
    pub last_local_save: Option<String>,
    pub is_user_logged_in: Option<bool>,
}

/// 用户设置
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct UserSettings {
    pub theme: Option<String>,
    pub language: Option<String>,
    pub autosave: Option<bool>,
    pub spell_check: Option<bool>,
    pub last_location: Option<bool>,
    pub font_size: Option<String>,
}

// ============================================================================
// 请求 DTO
// ============================================================================

/// 创建用户请求
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateUserRequest {
    pub username: String,
    pub display_name: Option<String>,
    pub avatar: Option<String>,
    pub email: Option<String>,
    pub plan: Option<UserPlan>,
    pub features: Option<UserFeatures>,
    pub settings: Option<UserSettings>,
}

/// 更新用户请求
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateUserRequest {
    pub username: Option<String>,
    pub display_name: Option<Option<String>>,
    pub avatar: Option<Option<String>>,
    pub email: Option<Option<String>>,
    pub last_login: Option<i64>,
    pub plan: Option<UserPlan>,
    pub plan_start_date: Option<Option<i64>>,
    pub plan_expires_at: Option<Option<i64>>,
    pub trial_expires_at: Option<Option<i64>>,
    pub token: Option<Option<String>>,
    pub server_message: Option<Option<String>>,
    pub features: Option<Option<UserFeatures>>,
    pub state: Option<Option<UserState>>,
    pub settings: Option<Option<UserSettings>>,
}

// ============================================================================
// 响应 DTO
// ============================================================================

/// 用户响应
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct UserResponse {
    pub id: String,
    pub username: String,
    pub display_name: Option<String>,
    pub avatar: Option<String>,
    pub email: Option<String>,
    pub last_login: i64,
    pub created_at: i64,
    pub plan: UserPlan,
    pub plan_start_date: Option<i64>,
    pub plan_expires_at: Option<i64>,
    pub trial_expires_at: Option<i64>,
    pub token: Option<String>,
    pub server_message: Option<String>,
    pub features: Option<UserFeatures>,
    pub state: Option<UserState>,
    pub settings: Option<UserSettings>,
}

/// Entity -> DTO 转换
impl From<super::user_entity::Model> for UserResponse {
    fn from(model: super::user_entity::Model) -> Self {
        Self {
            id: model.id,
            username: model.username,
            display_name: model.display_name,
            avatar: model.avatar,
            email: model.email,
            last_login: model.last_login,
            created_at: model.created_at,
            plan: model.plan,
            plan_start_date: model.plan_start_date,
            plan_expires_at: model.plan_expires_at,
            trial_expires_at: model.trial_expires_at,
            token: model.token,
            server_message: model.server_message,
            features: model.features.and_then(|s| serde_json::from_str(&s).ok()),
            state: model.state.and_then(|s| serde_json::from_str(&s).ok()),
            settings: model.settings.and_then(|s| serde_json::from_str(&s).ok()),
        }
    }
}
